### PR TITLE
[ECO-4987] Implement lifecycle behaviour that relates to an ongoing operation

### LINKED ISSUE #52: Implement lifecycle behaviour that relates to an ongoing operation
Split from #28.

I’m referring to behaviour described inside the specification of the four Room Lifecycle Operations (`ATTACH`, `DETACH`, `RELEASE`, `RETRY`), which describes what to do if another room lifecycle operation is in progress.

At the time of writing, with reference to spec version b4a495e, having so far looked at `ATTACH`, `DETACH`, and `RELEASE`, these are:

- [x] CHA-RL1d — states that an `ATTACH` should wait for current operation to complete
- [x] CHA-RL3c — states that `RELEASE` on a `RELEASING` room should return the result of the pending `RELEASE`

Update: Since creating this issue, I have also implemented the Room Lifecycle Monitoring part of the spec (CHA-RL4) for #53. Many of these spec points are predicated on whether there is a room lifecycle operation in progress. So, in order to not be blocked on implementing those spec points, I’ve introduced a `RoomLifecycleMonitor.hasOperationInProgress` boolean property that can be set by the tests. For the current task, we need to:

- [x] remove the ability to directly set this property and instead link it to whether there actually is a room lifecycle operation in progress

┆Issue is synchronized with this [Jira Task](https://ably.atlassian.net/browse/ECO-4987) by [Unito](https://www.unito.io)

### PR DESCRIPTION
**Pull request by lawrence-forooghian**
Does a bunch of groundwork, and then implements spec points CHA-RL1d and CHA-RL3c. See commit mesages for more details.

Resolves #52.

## Summary by CodeRabbit

- **New Features**
	- Enhanced room lifecycle management with a new status tracking system.
	- Improved handling of contributor states and operations, including attach, detach, and release.

- **Bug Fixes**
	- Refined logic for managing pending events and error handling during contributor detachment.

- **Tests**
	- Updated tests to reflect changes in room lifecycle operations and added new test cases for improved coverage and error handling.

### CODE REVIEW

File: Sources/AblyChat/RoomLifecycleManager.swift [hunk 1]
```diff
@@ -479,21 +685,38 @@ internal actor RoomLifecycleManager<Contributor: RoomLifecycleContributor> {
...
         case .detached:
             // CHA-RL3b
             changeStatus(to: .released)
             return
-        case .releasing, .initialized, .attached, .attaching, .detaching, .suspended, .failed:
+        case let .releasing(releaseOperationID):
+            // CHA-RL3c
+            // See note on waitForCompletionOfOperationWithID for the current need for this force try
+            // swiftlint:disable:next force_try
+            return try! await waitForCompletionOfOperationWithID(releaseOperationID, waitingOperationID: operationID)
```
[the remark this answers is not in the document]
Comment (lawrence-forooghian) [reply Sources/AblyChat/RoomLifecycleManager.swift:713]: there's a comment explaining why I've done this

File: Sources/AblyChat/RoomLifecycleManager.swift [hunk 2]
```diff
@@ -172,21 +135,131 @@ internal actor RoomLifecycleManager<Contributor: RoomLifecycleContributor> {
...
+            case .suspended,
+                 .initialized,
+                 .attached,
+                 .detached,
+                 .failed,
+                 .released:
+                nil
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (lawrence-forooghian) [reply Sources/AblyChat/RoomLifecycleManager.swift:191]: The cases _are_ all handled

File: Tests/AblyChatTests/RoomLifecycleManagerTests.swift [hunk 3]
```diff
@@ -125,6 +125,54 @@ struct RoomLifecycleManagerTests {
...
+        async let attachedWaitingForDetachedEvent = operationWaitEventSubscription.first { operationWaitEvent in
+            operationWaitEvent == .init(waitingOperationID: attachOperationID, waitedOperationID: detachOperationID)
+        }
+
+        // When: `performAttachOperation()` is called on the lifecycle manager
+        async let attachResult: Void = manager.performAttachOperation(testsOnly_forcingOperationID: attachOperationID)
+
+        // Then:
+        // - the manager informs us that the ATTACH operation is waiting for the DETACH operation to complete
+        // - when the DETACH completes, the ATTACH operation proceeds (which we check here by verifying that it eventually completes) — note that (as far as I can tell) there is no way to test that the ATTACH operation would have proceeded _only if_ the DETACH had completed; the best we can do is allow the manager to tell us that that this is indeed what it’s doing (which is what we check for in the previous bullet)
```
Comment (umair-ably) [inline Tests/AblyChatTests/RoomLifecycleManagerTests.swift:165]: I'm not sure I understand this... surely we _can_ test this to be the case by just waiting for a set time e.g. 10 secs _and_ not marking the detach operation as completed. In turn, we can check the attach operation does not complete in that time?

Comment (lawrence-forooghian) [reply Tests/AblyChatTests/RoomLifecycleManagerTests.swift:165]: 1. it just tells you that the attach operation didn't complete within 10 seconds; what if it were to complete within 11 seconds, say?
2. now you have a 10-second wait in your test (I know that your 10 seconds is arbitrary, but it's best to not have any waits if possible)

Comment (lawrence-forooghian) [reply Tests/AblyChatTests/RoomLifecycleManagerTests.swift:165]: > thinking out loud but could you not force attach to complete similar to how you have with contributorDetachOperation.complete(result: .success /* arbitrary */ )

No — the `contributorDetachOperation(…)` that you've mentioned forces a _contributor_ operation to complete, not a _room lifecycle operation_.

Comment (umair-ably) [reply Tests/AblyChatTests/RoomLifecycleManagerTests.swift:165]: ahh cool - makes sense

**Pull request merged**